### PR TITLE
db: fix overlap check for flushable ingest excises

### LINKED ISSUE #3963: internal/metamorphic: TestMeta failed [spurious key]
Encountered on a PR's CI run but it reproduces on current master SHA ([01dcf57](https://github.com/cockroachdb/pebble/commit/01dcf5752d56dfb7ded1255b3639cba6026d0fad)): https://github.com/cockroachdb/pebble/actions/runs/11040241548/job/30667767078?pr=3962

```
2024-09-25T20:08:06.3357754Z ===== SEED =====
2024-09-25T20:08:06.3358259Z 1727294659951576337
2024-09-25T20:08:06.3358854Z ===== DIFF =====
2024-09-25T20:08:06.3359960Z _meta/240925-200419.9512717887793/{standard-000,random-000}
2024-09-25T20:08:06.3360860Z @@ -3306,11 +3306,11 @@
2024-09-25T20:08:06.3361644Z  iter33.SeekGE("cpsifwvfka@3", "") // [false] <nil> #3305
2024-09-25T20:08:06.3362692Z  iter30.First() // [true,"cqqkguannt@6","",<no range>] <nil> #3306
2024-09-25T20:08:06.3364186Z  iter31.Last() // [true,"tvtwgzgsf",<no point>,["tvtwgzgsf","uxnufhrjlnpd")=>{"@6"="zbucyczbfyiknnhnzpwg"}] <nil> #3307
2024-09-25T20:08:06.3365662Z  db1.Set("ysaqsz@33", "vdxzpzjejpqq") // <nil> #3308
2024-09-25T20:08:06.3367390Z  iter27.SeekPrefixGE("ysaqsz@22") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #3309
2024-09-25T20:08:06.3368948Z -snap25.Get("mnwk@32") // [""] pebble: not found #3310
2024-09-25T20:08:06.3369909Z +snap25.Get("mnwk@32") // ["mnbtyjhmaaseexaqx"] <nil> #3310
2024-09-25T20:08:06.3374791Z  iter29.First() // [true,"cpsifwvfka@10",<no point>,["cpsifwvfka@10","dhvyv")=>{"@30"=""}*] <nil> #3311
2024-09-25T20:08:06.3376295Z  iter30.SeekPrefixGE("cqqkguannt@6") // [true,"cqqkguannt@6","",<no range>] <nil> #3312
2024-09-25T20:08:06.3377944Z  iter27.SetBounds("uxnufhrjlnpd@19", "ymogpw@8") // <nil> #3313
2024-09-25T20:08:06.3378741Z  iter27.SeekGE("uxnufhrjlnpd@19", "") // [true,"uxnufhrjlnpd@19","adhjmaaguwxjmijya",<no range>] <nil> #3314
2024-09-25T20:08:06.3379545Z  iter27.Prev("") // [false] <nil> #3315
```

Jira issue: PEBBLE-263

### PR DESCRIPTION
**Pull request by itsbilal**
Previously, we would incorrectly return false (denoting no overlap) between passed-in UserKeyBounds and a flushable ingest, if we see a file that starts after the passed-in bounds, even if there was an excise span overlapping with those bounds. This change fixes that, and adds a unit test to catch this edge case.

Fixes #3963.

### CODE REVIEW

**Review by jbowens — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3968#-:-O7jNcGi6toU2LwvdFoe:bnfp4nl)* status: 1 of 3 files reviewed, all discussions resolved (waiting on @RaduBerinde)

**Comment by itsbilal** [comment]
TFTR!

**Review by RaduBerinde — COMMENTED** [review]
Looks great, thanks for fixing so quickly!
